staging/src/k8s.io/kube-aggregator/pkg/util/dynamic_restconfig.go

```diff
@@ -0,0 +1,180 @@
+package util
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/server/dynamiccertificates"
+	"k8s.io/apiserver/pkg/server/egressselector"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const workItemKey = "key"
+
+type RestConfigProvider interface {
+	GetRestConfig() (*restclient.Config, error)
+	GetTransport() (http.RoundTripper, error)
+}
+
+// DynamicRestConfigProvider provides a restconfig and transport backed by dynamically loaded cert/key pair
+type DynamicRestConfigProvider struct {
+	certKeyPair       dynamiccertificates.CertKeyContentProvider
+	currentRestConfig atomic.Value
+	currentTransport  atomic.Value
+	queue             workqueue.RateLimitingInterface
+
+	// egressSelector selects the proper egress dialer to communicate with the custom apiserver
+	// overwrites proxyTransport dialer if not nil
+	egressSelector *egressselector.EgressSelector
+
+	// If present, the Dial method will be used for dialing out to delegate
+	// apiservers.
+	proxyTransport *http.Transport
+
+	insecure   bool
+	serverName string
+	caBundle   []byte
+}
+
+var _ dynamiccertificates.Listener = &DynamicRestConfigProvider{}
+
+func NewDynamicRestConfigProvider(servingContent dynamiccertificates.CertKeyContentProvider, egressSelector *egressselector.EgressSelector, proxyTransport *http.Transport, insecure bool, serverName string, caBundle []byte) (*DynamicRestConfigProvider, error) {
```

Comment:
it has to be dynamic, `serverName, caBundle` are not `constant`, what if they have changed?

Reply:
based on my understanding, these are provided when kube-aggregator starts so they cannot change without the process being restarted. But I might have missed something...

Reply:
check when `AddAPIService` function is called - https://github.com/kubernetes/kubernetes/pull/88120/files#r379316414